Automatically calculate and cache directive validations

skovati opened the pull request:
* **Tickets addressed:** closes https://github.com/NASA-AMMOS/aerie/issues/1195 closes #502 
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? -->
Implements automatic activity directive validation (plus caching of results in DB). This PR adds a new worker thread to `merlin-server` which queries `activity_directive_validations` for any directive that doesn't yet have its `validations` field populated.

The `activity_directive_validations` table is populated by database triggers on `activity_directive` insert / update, that will clear out existing validations for a given activity directive, causing the validation thread to pick up that work, and write fresh validation results.

For performance, the validation thread will poll for new validation requests every 500ms (configurable by env var), batch by mission model, thereby only loading each mission model once per poll. The database action that writes validation results back is also capable of batching writes.

The UI is now able to subscribe as follows to get validations for all activities in the currently selected plan, which it can use to populate the [validation error console](https://github.com/NASA-AMMOS/aerie-ui/issues/281)

```graphql
subscription directiveValidations {
  activity_directive_validations(
    where: {status: {_eq: "complete"}, 
    _and: {plan_id: {_eq: 1}}}
  ) {
    validations
    directive_id
  }
}

```

The `validations` JSON will have variable fields depending on what errors were encountered during validation. The JSON could be parsed roughly as follows:
```javascript
if (!validations.success) {
  // there are errors, need to dig deeper
  const { type, errors } = validations;
 
  // these top-level cases are mutually exclusive
  switch (type) {
    case: "NO_SUCH_ACTIVITY_TYPE":
      const { noSuchActivityType } = errors;
      // add to UI error list
      ...
      break;
    case: "VALIDATION_NOTICES":
      const { validationNotices } = errors;
      // add to UI error list
      ...
      break;
    case: "INSTANTIATION_ERRORS":
      // we have instantiation errors, which have sub error
      // types that are all simultaneously possible
      // (need to check size of all three remaining lists)
      const { missingArguments,
              extraneousArguments,
              unconstructableArguments } = errors;
      // add each list to UI error list
      ...
      break;
  }
}
```

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? -->
So far most of the testing has been performance based, utilizing load tests to ensure performance is sufficient. Manual testing so far has shown promising results, with a single request taking ~20ms:
```
[Thread-0] INFO gov.nasa.jpl.aerie.merlin.server.services.ValidationWorker - processing batch for mission model: 1
[Thread-0] INFO gov.nasa.jpl.aerie.merlin.server.services.ValidationWorker - processed model batch of size 1 in 17.452291 ms
```

... and batches of ~1000 requests taking ~60ms (once this validation code has moved up a few compilation tiers in the JVM JIT)
```
[Thread-0] INFO gov.nasa.jpl.aerie.merlin.server.services.ValidationWorker - processing batch for mission model: 1
[Thread-0] INFO gov.nasa.jpl.aerie.merlin.server.services.ValidationWorker - processed model batch of size 445 in 22.684292 ms
[Thread-0] INFO gov.nasa.jpl.aerie.merlin.server.services.ValidationWorker - processing batch for mission model: 2
[Thread-0] INFO gov.nasa.jpl.aerie.merlin.server.services.ValidationWorker - processed model batch of size 416 in 36.900708 ms
```

torture test of 10000 inserts on a single mission model is still quite quick
```
[Thread-0] INFO gov.nasa.jpl.aerie.merlin.server.services.ValidationWorker - processing batch for mission model: 2
[Thread-0] INFO gov.nasa.jpl.aerie.merlin.server.services.ValidationWorker - processed model batch of size 10000 in 134.382375 ms
```

Correctness tests can be added if desired.

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and completeness? -->
API docs need to be created, as querying this validation table might be useful to some API consumers.

## Future work
<!-- What next steps can we anticipate from here, if any? -->
Unblocks https://github.com/NASA-AMMOS/aerie-ui/issues/936.
Long term, it might make sense to spin this validation work off into it's own container, e.g. `merlin-validator` that can be independently scaled from `merlin-server`.

skovati committed:
add migrations for automatic directive validations

skovati committed:
hook up `ValidationWorker` to merlin-server driver

skovati committed:
add validation response serializers

skovati committed:
add bulk directive validation methods to `MissionModelRepository`

skovati committed:
add bulk directive validation DB actions

skovati committed:
enable validation thread in `merlin-server` with env var

skovati committed:
add `ValidationWorker` service

skovati committed:
add bulk directive validation methods to `MissionModelService`

skovati committed:
add sql for automatic directive validations

skovati commented on merlin-server/sql/merlin/tables/activity_directive.sql:
```diff
@@ -183,14 +183,22 @@ execute function activity_directive_set_updated_at();
   'Sets the last_modified_at field of an activity_directive to the current time.';
 
-create function activity_directive_set_arguments_updated_at()
+create or replace function activity_directive_set_arguments_updated_at()
```
Remove this `or replace`

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/MissionModelService.java:
```diff
@@ -107,4 +103,14 @@ record Success(SerializedActivity activity) implements  BulkEffectiveArgumentRes
+        BulkArgumentValidationResponse { }
+    record InstantiationError(InstantiationException ex) implements
+        BulkArgumentValidationResponse { }
+  }
```
🔵 Extremely minor - I like to make these one-liners when I can
```suggestion
  sealed interface BulkArgumentValidationResponse {
    record Success() implements BulkArgumentValidationResponse { }
    record Validation(List<ValidationNotice> notices) implements BulkArgumentValidationResponse { }
    record NoSuchActivityError(NoSuchActivityTypeException ex) implements BulkArgumentValidationResponse { }
    record InstantiationError(InstantiationException ex) implements BulkArgumentValidationResponse { }
  }
```

skovati replied:
intellij formatting 🙄, change has been made.

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/ResponseSerializers.java:
```diff
@@ -158,6 +159,35 @@ public static JsonValue serializeBulkEffectiveArgumentResponse(BulkEffectiveArgu
+                     .add("validationNotices", serializeIterable(ResponseSerializers::serializeValidationNotice, v.notices()))
+                     .build())
+                 .build();
+    } else if (response instanceof BulkArgumentValidationResponse.NoSuchActivityError e) {
+      return serializeNoSuchActivityTypeException(e.ex());
+    } else if (response instanceof BulkArgumentValidationResponse.InstantiationError f) {
+      return serializeInstantiationException(f.ex());
+    }
```
🔶 From a UI parsing perspective, it would be great if the structure of the object was consistent across all cases - I'm not 100% sure what that would look like, but I'll point out some inconsistencies here. Here's a quick survey of the current format:

```js
// Success
{
  "success": true
}
// Instantiation failure
{
  "success": false,
  "errors": {
    "missingArguments": [],
    "extraneousArguments": ["quanitid"],
    "unconstructableArguments": []
   },
  "arguments": {"quantity": 1, "growingDuration": 3600000000}
}
// Validation failure
{
  "success": false,
  "errors": {
    "validationNotices": [{"message": "Quantity must be positive", "subjects": ["quantity"]}]
  }
}
// No such activity type
{
  "message": "no such activity type",
  "activity_type": "GrowlBanana"
}
```

Some inconsistencies:
- The "no such activity type" response lacks a `success` property, and in the `"success": false` case it lacks an `errors` property
- While both instantiation failures and validation failures share an `errors` property, the structure of the `errors` object differs, and there's no way to figure out which is which except by checking for the presence or absence of fields. Perhaps we could add `"type": "INSTANTIATION_FAILURE"` and `"type": "VALIDATION_FAILURE"` respectively
- `"arguments"` shows up in just one of the four types

skovati replied:
Agreed. I originally wanted to reuse the existing `ResponseSerializers` functions, but unifying the response schema is worth adding some new serializers.

skovati replied:
I've made a small change just to the "no such activity type" case that puts this error under an `errors` field like the other responses. I figured that the UI will either have to switch on fields being null or `type === "SOME_STRING"`, but this way we'll be storing (marginally) less data in the validation results for each directive in the DB.

skovati replied:
After chatting with the UI folks, they agreed that switching on a string v.s. nullity of fields is easier, so I've added a enum style string to the schema for each of these error results.

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/http/ResponseSerializers.java:
```diff
@@ -158,6 +159,35 @@ public static JsonValue serializeBulkEffectiveArgumentResponse(BulkEffectiveArgu
+
+    return Json.createObjectBuilder()
+               .add("success", JsonValue.FALSE)
+               .add("errors", String.format("Internal error: %s", response))
```
🔵 I'd be okay with treating this code as "this should never happen" and leaving it out of the consistent error format discussion. Feel free to spruce it up if you'd like, though

skovati replied:
I'll leave this as is for now, since Java doesn't know this case can't happen and needs it to compile.

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/GetUnvalidatedDirectivesAction.java:
```diff
@@ -0,0 +1,68 @@
+      final var arguments = getJsonColumn(results, "arguments", activityArgumentsP)
+          .getSuccessOrThrow($ -> new Error("Corrupt activity arguments cannot be parsed: " + $.reason()));
+
+      if (!map.containsKey(modelId)) {
+        map.put(modelId, new ArrayList<>());
+      }
+
+      map.get(modelId)
+         .add(new ActivityDirectiveForValidation(
+             new ActivityDirectiveId(directiveId),
+             new PlanId(planId),
+             new SerializedActivity(type, arguments)
+         ));
```
🔵 As minor as it gets
```suggestion
      map.computeIfAbsent(modelId, $ -> new ArrayList<>())
         .add(new ActivityDirectiveForValidation(
             new ActivityDirectiveId(directiveId),
             new PlanId(planId),
             new SerializedActivity(type, arguments)
         ));
```

skovati replied:
Clean. Change has been made.

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresMissionModelRepository.java:
```diff
@@ -146,20 +146,29 @@ public void updateResourceTypes(final String missionModelId, final Map<String, R
+          updateAction.addBatch(directive.id(), directive.planId(), validation);
+        }
+        updateAction.executeBatch();
+      }
```
🔵 These two "try" statements can be combined into one
```suggestion
    try (final var connection = this.dataSource.getConnection();
         final var updateAction = new UpdateActivityDirectiveValidationsAction(connection)) {
      for (final var update : updates) {
        final var directive = update.getLeft();
        final var validation = update.getRight();
        updateAction.addBatch(directive.id(), directive.planId(), validation);
      }
      updateAction.executeBatch();
```

skovati replied:
Change has been made to both functions added in this file.

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/UpdateActivityDirectiveValidationsAction.java:
```diff
@@ -27,17 +22,30 @@ public UpdateActivityDirectiveValidationsAction(final Connection connection) thr
   }
 
-  public void apply(final long directiveId, final long planId, final Timestamp argumentsModifiedTime, final List<ValidationNotice> notices)
+  public void apply(final ActivityDirectiveId directiveId, final PlanId planId, BulkArgumentValidationResponse results)
```
🔶 If I understand correctly, this method is now unused? If so, we should consider deleting it.

🔵 Alternatively, `apply` could take in the whole list of updates, and have a for loop for updating the batch - this would be a bit more in line with some other actions (e.g. see `InsertActivityTypes`)

skovati replied:
I debated passing a list in, but figured it might be useful to keep around a version that acted on a single validation. Thinking about it more, I suppose the bulk version can just be passed a list of size 1 if we did want this functionality. Change has been made.

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ValidationWorker.java:
```diff
@@ -0,0 +1,58 @@
+    logger.info("validation worker starting...");
+    while (!Thread.interrupted()) {
+      try {
+        Thread.sleep(500);
```
🔶 Since it doesn't hurt, let's make this a configurable environment variable

skovati replied:
Change has been made.

mattdailis replied:
There's a file called `Environment.md` - let's make sure the two new environment variables are documented there

skovati replied:
Updated!

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ValidationWorker.java:
```diff
@@ -0,0 +1,58 @@
+        Thread.sleep(500);
+
+        final var validationRequests = missionModelService.getUnvalidatedDirectives();
+        logger.info(
```
🔶 The log feels a little noisy - maybe consider logging only when the number of validations is greater than zero? Or moving the log level to `debug`

skovati replied:
Good call, I've implemented a conditional check to only log when requests are `> 0`, and moved these log statements to `debug`.

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/ValidationWorker.java:
```diff
@@ -0,0 +1,58 @@
+        }
+
+      } catch (NoSuchMissionModelException ex) {
+        throw new RuntimeException("Validation request failed due to missing mission model", ex);
```
🔶 I'm curious about what happens when this exception is thrown (or if any other uncaught exception occurs while checking validations). Does the thread die? I'd like to dig into our fault behavior here a little bit to make sure we understand it

skovati replied:
Good question. On second thought, I've changed this case to an `error` log instead, although I can't see a case where this exception actually gets thrown, due to DB foreign key constraints.

Mythicaeda replied:
> I can't see a case where this exception actually gets thrown, due to DB foreign key constraints

`null` is permitted for a plan's `model_id`. In fact, if you delete a mission model, all plans using that model will automatically have their `model_id` set to `null`.

skovati replied:
Furthermore, I've added a `catch (Throwable)` to this error handling, which should keep this validation thread running regardless of exceptions thrown.

mattdailis reviewed (commented):
> Key:
> 🔻 Issues to address before merge
>  🔶 Requests that should not block merge, but should at least be discussed 
> 🔵 Recommendations that can be ignored if desired

skovati committed:
use new explicit `status` field as validation trigger

skovati committed:
Add error type to validation JSON schema

skovati committed:
rollback transaction if any validation updates fail

skovati committed:
add non-fatal catch all handler to validation worker thread

skovati committed:
add new validation thread env vars to docs

skovati committed:
refactor `ValidationWorker` for clarity

skovati committed:
update `ActivityDirectiveForValidation` with argument timestamp

This is done to avoid the following race condition:

- merlin-server queries for unvalidated directives
- merlin-server starts validation process
- aerie-ui updates directive that merlin-server is currently processing
- merlin-server writes validation and status = complete
- merlin-server doesn't pick up new directive arguments on next query, since status = complete

This change will make it so the validation worker thread only writes status = complete if the
arguments last modified at timestamp matches, meaning a new request didn't come in during validation

skovati committed:
update model records for validation work

skovati committed:
add `ExceptionActivity` to banananation

mattdailis committed:
Catch throwable in getValidationFailures

skovati committed:
add initial e2e tests for automatic activity validations

skovati committed:
test all error cases of directive validation responses
